pubsublite/internal/test/msg_validators.go

```diff
@@ -0,0 +1,172 @@
+// stream breaks while there are in-flight batches, which are resent upon
+// stream reconnect.
+func (or *OrderingReceiver) Receive(data, key string) error {
+	or.mu.Lock()
+	defer or.mu.Unlock()
+
+	idx := parseMsgIndex(data)
+	if idx < 0 {
+		return fmt.Errorf("failed to parse index from message: %q", data)
+	}
+
+	// Verify non-decreasing ordering. Allow duplicates, which can be verified
+	// with DuplicateMsgDetector.
+	lastIdx, exists := or.received[key]
+	if exists && idx < lastIdx {
+		return fmt.Errorf("message ordering failed for key %s, expected message idx >= %d, got %d", key, lastIdx, idx)
+	}
+	or.received[key] = idx
+	return nil
+}
+
+var void struct{}
+
+type msgMetadata struct {
+	offsets map[int64]struct{}
+}
+
+func newMsgMetadata() *msgMetadata {
+	return &msgMetadata{
+		offsets: make(map[int64]struct{}),
+	}
+}
+
+func (mm *msgMetadata) ContainsOffset(offset int64) bool {
+	_, exists := mm.offsets[offset]
+	return exists
+}
+
+func (mm *msgMetadata) AddOffset(offset int64) {
+	mm.offsets[offset] = void
+}
+
+// DuplicateMsgDetector can be used to detect duplicate messages, either due to
+// duplicate publishes or receives.
+type DuplicateMsgDetector struct {
+	mu sync.Mutex
+	// Map of Pub/Sub message data and associated metadata.
+	msgs                  map[string]*msgMetadata
+	duplicatePublishCount int64
+	duplicateReceiveCount int64
+}
+
+// NewDuplicateMsgDetector creates a new DuplicateMsgDetector.
+func NewDuplicateMsgDetector() *DuplicateMsgDetector {
+	return &DuplicateMsgDetector{
+		msgs: make(map[string]*msgMetadata),
+	}
+}
+
+// Receive checks the given message data and offset.
+func (dm *DuplicateMsgDetector) Receive(data string, offset int64) {
+	dm.mu.Lock()
+	defer dm.mu.Unlock()
+
+	if metadata, exists := dm.msgs[data]; exists {
+		if metadata.ContainsOffset(offset) {
+			// If the message contains the same offset, it means it was received
+			// multiple times. This is not expected within a single test run. But it
+			// is normal when processes are stopped & restarted without committing
+			// cursors.
+			dm.duplicateReceiveCount++
+		} else {
+			// If the message contains a different offset, it means a message was
+			// republished, which can occur when a publish stream reconnects with
+			// in-flight published messages.
+			dm.duplicatePublishCount++
+			metadata.AddOffset(offset)
+		}
+	} else {
+		metadata = newMsgMetadata()
+		metadata.AddOffset(offset)
+		dm.msgs[data] = metadata
+	}
+}
+
+// Status returns a non-empty status string if there were duplicates detected.
+func (dm *DuplicateMsgDetector) Status() string {
+	dm.mu.Lock()
+	defer dm.mu.Unlock()
+
+	if (dm.duplicateReceiveCount + dm.duplicatePublishCount) == 0 {
+		return ""
+	}
+	return fmt.Sprintf("duplicate publish count = %d, receive count = %d", dm.duplicatePublishCount, dm.duplicateReceiveCount)
+}
+
+// HasDuplicates returns true if duplicate messages were detected.
+func (dm *DuplicateMsgDetector) HasDuplicates() bool {
```

Comment:
Should we split this into HasReceiveDuplicates and HasPublishDuplicates? It sounds to me that publish duplicates are expected in normal operation, but receive duplicates are indicative of a serious bug.

Reply:
Good point. Done.